File: plugins/hermes-plugin/src/tools.ts
import { z } from 'zod';

const toolNameSchema = z.enum([
  'web_search_to_url',
  'youtube_search_to_url',
  'canvas_set_page',
  'canvas_navigate_panel',
  'canvas_media_play',
  'canvas_media_control',
]);

const executeSchema = z.object({
  tool: toolNameSchema,
  input: z.record(z.unknown()).default({}),
});

export type ToolName = z.infer<typeof toolNameSchema>;
export type ExecuteRequest = z.infer<typeof executeSchema>;

interface ToolResult {
  ok: boolean;
  tool: ToolName;
  result?: Record<string, unknown>;
  error?: string;
}

const canvasApiUrl = (process.env.CANVAS_API_URL ?? 'http://127.0.0.1:3100').replace(/\/$/, '');

function decodeHtmlEntities(value: string): string {
  return value
    .replace(/&amp;/g, '&')
    .replace(/&quot;/g, '"')
    .replace(/&#x27;/g, "'")
    .replace(/&#39;/g, "'")
    .replace(/&lt;/g, '<')
    .replace(/&gt;/g, '>');
}

function looksLikeWebUrl(value: string): boolean {
  try {
    const u = new URL(value);
    return ['http:', 'https:'].includes(u.protocol);
  } catch {
    return false;
  }
}

function normalizeUrl(value: string): string {
  const trimmed = value.trim();
  if (!trimmed) return '';
  if (/^https?:\/\//i.test(trimmed)) return trimmed;
  return `https://${trimmed}`;
}

const mediaSourceSchema = z.enum([
  'music_assistant',
  'radio_browser',
  'direct_audio',
  'youtube',
]);

const mediaActionSchema = z.enum(['pause', 'resume', 'stop', 'volume', 'mute']);

async function resolveWebQueryToUrl(query: string): Promise<string> {
  const trimmed = query.trim();
  if (!trimmed) return '';

  const searchEndpoints = [
    `https://duckduckgo.com/html/?q=${encodeURIComponent(trimmed)}`,
    `https://lite.duckduckgo.com/lite/?q=${encodeURIComponent(trimmed)}`,
  ];

  for (const endpoint of searchEndpoints) {
    try {
      const response = await fetch(endpoint, {
        headers: {
          'User-Agent': 'Mozilla/5.0 (X11; Linux x86_64) CanvasDisplay-HermesPlugin/1.0',
          'Accept-Language': 'en-US,en;q=0.9',
        },
      });
      if (!response.ok) continue;
      const html = await response.text();
      const linkPattern = new RegExp('href="([^"]*(?:uddg=|https?:\\/\\/)[^"]+)"', 'gi');
      const redirectMatches = Array.from(html.matchAll(linkPattern));

      for (const match of redirectMatches) {
        const href = decodeHtmlEntities(match[1] ?? '');
        if (!href) continue;

        if (href.includes('uddg=')) {
          try {
            const parsedHref = new URL(href, endpoint);
            const uddg = parsedHref.searchParams.get('uddg');
            if (uddg && looksLikeWebUrl(uddg)) return uddg;
          } catch {
            // ignore malformed redirect links
          }
        }

        if (looksLikeWebUrl(href)) return href;
      }
    } catch {
      // try next endpoint
    }
  }

  return `https://duckduckgo.com/?q=${encodeURIComponent(trimmed)}`;
}

async function callCanvas(path: string, body: Record<string, unknown>): Promise<Record<string, unknown>> {
  const response = await fetch(`${canvasApiUrl}${path}`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  });

  const text = await response.text();
  let parsed: unknown = null;
  try {
    parsed = text ? JSON.parse(text) : null;
  } catch {
    parsed = { raw: text };
  }

  if (!response.ok) {
    throw new Error(`Canvas API ${path} failed (${response.status}): ${text}`);
  }

  return (parsed && typeof parsed === 'object') ? parsed as Record<string, unknown> : { value: parsed };
}

export function getToolManifest(): Array<Record<string, unknown>> {
  return [
    {
      name: 'web_search_to_url',
      description: 'Resolve a web query to a reliable URL with search fallback.',
      inputSchema: {
        type: 'object',
        properties: {
          query: { type: 'string' },
        },
        required: ['query'],
      },
    },
    {
      name: 'youtube_search_to_url',
      description: 'Create a YouTube search URL from query text.',
      inputSchema: {
        type: 'object',
        properties: {
          query: { type: 'string' },
        },
        required: ['query'],
      },
    },
    {
      name: 'canvas_set_page',
      description: 'Set active Canvas page by page_id or page name.',
      inputSchema: {
        type: 'object',
        properties: {
          page_id: { type: 'string' },
          page: { type: 'string' },
        },
      },
    },
    {
      name: 'canvas_navigate_panel',
      description: 'Navigate a Canvas panel to a URL by panel_id or panel name.',
      inputSchema: {
        type: 'object',
        properties: {
          panel_id: { type: 'string' },
          panel: { type: 'string' },
          page_id: { type: 'string' },
          page: { type: 'string' },
          url: { type: 'string' },
        },
        required: ['url'],
      },
    },
    {
      name: 'canvas_media_play',
      description: 'Play media via Canvas unified media API.',
      inputSchema: {
        type: 'object',
        properties: {
          source: {
            type: 'string',
            enum: ['music_assistant', 'radio_browser', 'direct_audio', 'youtube'],
          },
          url: { type: 'string' },
          title: { type: 'string' },
          volume: { type: 'number' },
          panel_id: { type: 'string' },
        },
        required: ['source', 'url'],
      },
    },
    {
      name: 'canvas_media_control',
      description: 'Control media playback via Canvas unified media API.',
      inputSchema: {
        type: 'object',
        properties: {
          source: {
            type: 'string',
            enum: ['music_assistant', 'radio_browser', 'direct_audio', 'youtube'],
          },
          action: {
            type: 'string',
            enum: ['pause', 'resume', 'stop', 'volume', 'mute'],
          },
          level: { type: 'number' },
          muted: { type: 'boolean' },
        },
        required: ['source', 'action'],
      },
    },
  ];
}

export async function executeTool(raw: unknown): Promise<ToolResult> {
  const parsed = executeSchema.safeParse(raw);
  if (!parsed.success) {
    return {
      ok: false,
      tool: 'web_search_to_url',
      error: parsed.error.flatten().formErrors.join('; ') || 'Invalid execute payload',
    };
  }

  const { tool, input } = parsed.data;

  try {
    if (tool === 'web_search_to_url') {
      const query = String(input.query ?? '').trim();
      if (!query) throw new Error('query is required');
      const url = await resolveWebQueryToUrl(query);
      return { ok: true, tool, result: { query, url } };
    }

    if (tool === 'youtube_search_to_url') {
      const query = String(input.query ?? '').trim();
      if (!query) throw new Error('query is required');
      const url = `https://www.youtube.com/results?search_query=${encodeURIComponent(query)}`;
      return { ok: true, tool, result: { query, url } };
    }

    if (tool === 'canvas_set_page') {
      const page_id = typeof input.page_id === 'string' ? input.page_id.trim() : '';
      const page = typeof input.page === 'string' ? input.page.trim() : '';
      if (!page_id && !page) throw new Error('page_id or page is required');
      const result = await callCanvas('/api/commands/page', {
        ...(page_id ? { page_id } : {}),
        ...(page ? { page } : {}),
      });
      return { ok: true, tool, result };
    }

    if (tool === 'canvas_navigate_panel') {
      const panel_id = typeof input.panel_id === 'string' ? input.panel_id.trim() : '';
      const panel = typeof input.panel === 'string' ? input.panel.trim() : '';
      const page_id = typeof input.page_id === 'string' ? input.page_id.trim() : '';
      const page = typeof input.page === 'string' ? input.page.trim() : '';
      const rawUrl = typeof input.url === 'string' ? input.url.trim() : '';
      if (!rawUrl) throw new Error('url is required');
      const url = normalizeUrl(rawUrl);
      const result = await callCanvas('/api/commands/navigate', {
        ...(panel_id ? { panel_id } : {}),
        ...(panel ? { panel } : {}),
        ...(page_id ? { page_id } : {}),
        ...(page ? { page } : {}),
        url,
      });
      return { ok: true, tool, result: { ...result, url } };
    }

    if (tool === 'canvas_media_play') {
      const source = mediaSourceSchema.parse(String(input.source ?? '').trim());
      const rawUrl = String(input.url ?? '').trim();
      if (!rawUrl) throw new Error('url is required');

      const title = typeof input.title === 'string' ? input.title.trim() : '';
      const panel_id = typeof input.panel_id === 'string' ? input.panel_id.trim() : '';
      const volume = typeof input.volume === 'number' ? input.volume : undefined;

      const result = await callCanvas('/api/media/play', {
        source,
        url: source === 'youtube' ? rawUrl : normalizeUrl(rawUrl),
        ...(title ? { title } : {}),
        ...(panel_id ? { panel_id } : {}),
        ...(volume !== undefined ? { volume } : {}),
      });

      return { ok: true, tool, result };
    }

    if (tool === 'canvas_media_control') {
      const source = mediaSourceSchema.parse(String(input.source ?? '').trim());
      const action = mediaActionSchema.parse(String(input.action ?? '').trim());

      const payload: Record<string, unknown> = { source, action };
      if (typeof input.level === 'number') payload.level = input.level;
      if (typeof input.muted === 'boolean') payload.muted = input.muted;

      const result = await callCanvas('/api/media/control', payload);
      return { ok: true, tool, result };
    }

    return { ok: false, tool, error: `Unsupported tool: ${tool}` };
  } catch (err) {
    return { ok: false, tool, error: (err as Error).message };
  }
}

File: plugins/hermes-plugin/tools.py
"""Hermes tool handlers for Canvas Display plugin.

Each handler returns a JSON string and accepts (args, **kwargs).
"""

from __future__ import annotations

import json
import os
import re
from html import unescape
from urllib.error import HTTPError, URLError
from urllib.parse import parse_qs, quote_plus, urljoin, urlparse
from urllib.request import Request, urlopen

CANVAS_API_URL = os.environ.get("CANVAS_API_URL", "http://127.0.0.1:3100").rstrip("/")


def _json(payload: dict) -> str:
    return json.dumps(payload)


def _looks_like_web_url(value: str) -> bool:
    try:
        parsed = urlparse(value)
        return parsed.scheme in {"http", "https"} and bool(parsed.netloc)
    except Exception:
        return False


def _normalize_url(value: str) -> str:
    trimmed = value.strip()
    if not trimmed:
        return ""
    if re.match(r"^https?://", trimmed, flags=re.IGNORECASE):
        return trimmed
    return f"https://{trimmed}"


def _read_text(url: str) -> str:
    req = Request(
        url,
        headers={
            "User-Agent": "Mozilla/5.0 (X11; Linux x86_64) CanvasDisplay-HermesPlugin/1.0",
            "Accept-Language": "en-US,en;q=0.9",
        },
    )
    with urlopen(req, timeout=10) as resp:
        return resp.read().decode("utf-8", errors="replace")


def _resolve_web_query_to_url(query: str) -> str:
    trimmed = query.strip()
    if not trimmed:
        return ""

    endpoints = [
        f"https://duckduckgo.com/html/?q={quote_plus(trimmed)}",
        f"https://lite.duckduckgo.com/lite/?q={quote_plus(trimmed)}",
    ]

    link_pattern = re.compile(r'href="([^"]*(?:uddg=|https?://)[^"]+)"', re.IGNORECASE)

    for endpoint in endpoints:
        try:
            html = _read_text(endpoint)
        except Exception:
            continue

        for match in link_pattern.finditer(html):
            href = unescape(match.group(1) or "")
            if not href:
                continue

            if "uddg=" in href:
                try:
                    parsed_href = urlparse(urljoin(endpoint, href))
                    uddg_values = parse_qs(parsed_href.query).get("uddg")
                    if uddg_values:
                        uddg = uddg_values[0]
                        if _looks_like_web_url(uddg):
                            return uddg
                except Exception:
                    pass

            if _looks_like_web_url(href):
                return href

    return f"https://duckduckgo.com/?q={quote_plus(trimmed)}"


def _call_canvas(path: str, body: dict) -> dict:
    url = f"{CANVAS_API_URL}{path}"
    data = json.dumps(body).encode("utf-8")
    req = Request(url, data=data, method="POST")
    req.add_header("Content-Type", "application/json")

    try:
        with urlopen(req, timeout=10) as resp:
            raw = resp.read().decode("utf-8", errors="replace")
            return json.loads(raw) if raw else {}
    except HTTPError as err:
        raw = err.read().decode("utf-8", errors="replace") if err.fp else ""
        raise RuntimeError(f"Canvas API {path} failed ({err.code}): {raw}") from err
    except URLError as err:
        raise RuntimeError(f"Canvas API {path} request failed: {err.reason}") from err


def web_search_to_url(args: dict, **kwargs) -> str:
    try:
        query = str(args.get("query", "")).strip()
        if not query:
            return _json({"error": "query is required"})
        url = _resolve_web_query_to_url(query)
        return _json({"query": query, "url": url})
    except Exception as exc:
        return _json({"error": str(exc)})


def youtube_search_to_url(args: dict, **kwargs) -> str:
    try:
        query = str(args.get("query", "")).strip()
        if not query:
            return _json({"error": "query is required"})
        url = f"https://www.youtube.com/results?search_query={quote_plus(query)}"
        return _json({"query": query, "url": url})
    except Exception as exc:
        return _json({"error": str(exc)})


def canvas_set_page(args: dict, **kwargs) -> str:
    try:
        page_id = str(args.get("page_id", "")).strip()
        page = str(args.get("page", "")).strip()
        if not page_id and not page:
            return _json({"error": "page_id or page is required"})

        payload = {}
        if page_id:
            payload["page_id"] = page_id
        if page:
            payload["page"] = page

        result = _call_canvas("/api/commands/page", payload)
        return _json(result)
    except Exception as exc:
        return _json({"error": str(exc)})


def canvas_navigate_panel(args: dict, **kwargs) -> str:
    try:
        panel_id = str(args.get("panel_id", "")).strip()
        panel = str(args.get("panel", "")).strip()
        page_id = str(args.get("page_id", "")).strip()
        page = str(args.get("page", "")).strip()
        raw_url = str(args.get("url", "")).strip()
        if not raw_url:
            return _json({"error": "url is required"})

        url = _normalize_url(raw_url)
        payload = {"url": url}
        if panel_id:
            payload["panel_id"] = panel_id
        if panel:
            payload["panel"] = panel
        if page_id:
            payload["page_id"] = page_id
        if page:
            payload["page"] = page

        result = _call_canvas("/api/commands/navigate", payload)
        result["url"] = url
        return _json(result)
    except Exception as exc:
        return _json({"error": str(exc)})


def canvas_media_play(args: dict, **kwargs) -> str:
    try:
        source = str(args.get("source", "")).strip()
        if not source:
            return _json({"error": "source is required"})

        url = str(args.get("url", "")).strip()
        if not url:
            return _json({"error": "url is required"})

        title = str(args.get("title", "")).strip()
        panel_id = str(args.get("panel_id", "")).strip()
        volume = args.get("volume", None)

        payload = {
            "source": source,
            "url": url,
        }
        if title:
            payload["title"] = title
        if panel_id:
            payload["panel_id"] = panel_id
        if volume is not None:
            payload["volume"] = volume

        result = _call_canvas("/api/media/play", payload)
        return _json(result)
    except Exception as exc:
        return _json({"error": str(exc)})


def canvas_media_control(args: dict, **kwargs) -> str:
    try:
        source = str(args.get("source", "")).strip()
        if not source:
            return _json({"error": "source is required"})

        action = str(args.get("action", "")).strip()
        if not action:
            return _json({"error": "action is required"})

        payload = {
            "source": source,
            "action": action,
        }

        if "level" in args:
            payload["level"] = args.get("level")
        if "muted" in args:
            payload["muted"] = args.get("muted")

        result = _call_canvas("/api/media/control", payload)
        return _json(result)
    except Exception as exc:
        return _json({"error": str(exc)})
